pyexchange/fix.py

```diff
@@ -152,32 +185,95 @@ def write(self, message: simplefix.FixMessage):
         self.write_queue.put(message)
         pass
 
-    async def _wait_for_response(self, message_type: str) -> simplefix.FixMessage:
+    def _get_client_id(self, order_id: str) -> str:
+        """ Retrieve the client_id from an order_id structured as exchange_id|client_id """
+        assert isinstance(order_id, str)
+
+        return order_id.split('|')[1]
+
+    # TODO: return a List[{order_id: amount_remaining}]
+    async def _sync_orders(self, orders: List[Order]) -> List[dict]:
+        """
+            Iterate through clients list of orders and update according to exchange state.
+
+            If an order has been canceled or completely filled, delete the entry in the order book.
+
+            Finally, return a list of dictionaries: {order_id: filled_amount}.
+            This enables us to keep amounts in sync in the event of partial fills.
+        """
+        assert isinstance(orders, List)
+
+        cancel_message_types = [simplefix.EXECTYPE_CANCELED, simplefix.EXECTYPE_EXPIRED]
+        open_orders = []
+
+        for order in orders:
+            client_order_id = self._get_client_id(order.order_id)
+            order_status_dict = {}
+
+            while not self.order_book[client_order_id].empty():
+                exchange_order_state = self.order_book[client_order_id].get()
```

Comment:
Here's a list of possible states and my commentary on them:

```
EXECTYPE_NEW = b'0'		- how are these handled?
EXECTYPE_PARTIAL_FILL = b'1'	- handled
EXECTYPE_FILL = b'2'		- handled
EXECTYPE_DONE_FOR_DAY = b'3'	- needed for GTD orders
EXECTYPE_CANCELED = b'4'	- handled
EXECTYPE_REPLACE = b'5'		- probably not needed since we don't implement cxl/replace
EXECTYPE_PENDING_CANCEL = b'6'  - how are these handled?
EXECTYPE_STOPPED = b'7'		- unsure
EXECTYPE_REJECTED = b'8'	- should handle invalid orders
EXECTYPE_SUSPENDED = b'9'	- probably not needed
EXECTYPE_PENDING_NEW = b'A'	- how are these handled?
EXECTYPE_CALCULATED = b'B'	- probably not needed
EXECTYPE_EXPIRED = b'C'		- handled
EXECTYPE_RESTATED = b'D'	- probably not needed
EXECTYPE_PENDING_REPLACE = b'E' - probably not needed since we don't implement cxl/replace
```

Here's some useful documentation regarding the state machine: 
https://www.fixtrading.org/online-specification/order-state-changes/

Just want to ensure we're handling all relevant states.

Reply:
Thanks, this is very helpful. As you pointed out, pending messages aren't currently handled - this is a byproduct of the ErisX design in that they don't seem to send `PENDING_CANCEL` or `PENDING_NEW` messages. Additionally, some orders can be rejected with `35=8` and non-standard `103=100` or `103=23` codes. I just added additional checks for these as well.

Reply:
'EXECTYPE_NEW' is handled by the catchall message return on L269 of `_wait_for_order_processing_response()`. `EXECTYPE_DONE_FOR_DAY` is not currently supported by ErisX. In the event that an Order is cancelled by ErisX, it will be recorded as a standard Order Cancellation Exectuion Report, which sync_orders should catch and remove from our client's order book.